File: ios/TabletHID/Store/SessionLogger.swift
import Foundation

final class SessionLogger {
    private let writer: FileHandle
    private let queue = DispatchQueue(label: "hid-logger", qos: .utility)
    private var closed = false

    private let isoFmt: DateFormatter = {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        f.locale = Locale(identifier: "en_US_POSIX")
        return f
    }()

    init(mode: DeviceMode, profileName: String, touchConfig: TouchMouseConfig?) throws {
        let dir = Self.sessionDir()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)

        let stampFmt = DateFormatter()
        stampFmt.dateFormat = "yyyyMMdd_HHmmss"
        stampFmt.locale = Locale(identifier: "en_US_POSIX")
        let stem = "\(stampFmt.string(from: Date()))_\(mode.rawValue.lowercased())"

        let configStr = Self.buildConfig(mode: mode, profileName: profileName, touch: touchConfig)
        try configStr.write(to: dir.appendingPathComponent("\(stem).config"), atomically: true, encoding: .utf8)

        let logURL = dir.appendingPathComponent("\(stem).log")
        FileManager.default.createFile(atPath: logURL.path, contents: nil)
        writer = try FileHandle(forWritingTo: logURL)

        let ts = now()
        queue.async { [weak self] in
            self?.writeLine("\(ts) SESSION_START mode=\(mode.rawValue.uppercased())")
        }
    }

    func logMouse(buttons: Int, dx: Int, dy: Int, wheel: Int) {
        guard !closed else { return }
        let ts = now()
        queue.async { [weak self] in
            self?.writeLine("\(ts) MOUSE buttons=\(buttons) dx=\(dx) dy=\(dy) wheel=\(wheel)")
        }
    }

    func logGamepad(lx: Int, ly: Int, rx: Int, ry: Int, lt: Int, rt: Int, buttons: Int, hat: Int) {
        guard !closed else { return }
        let ts = now()
        queue.async { [weak self] in
            self?.writeLine("\(ts) GAMEPAD lx=\(lx) ly=\(ly) rx=\(rx) ry=\(ry) lt=\(lt) rt=\(rt) buttons=\(buttons) hat=\(hat)")
        }
    }

    func close() {
        guard !closed else { return }
        closed = true
        let ts = now()
        queue.async { [weak self] in
            guard let self else { return }
            writeLine("\(ts) SESSION_END")
            try? writer.synchronize()
            try? writer.close()
        }
    }

    private func writeLine(_ line: String) {
        guard let data = (line + "\n").data(using: .utf8) else { return }
        try? writer.write(contentsOf: data)
    }

    private func now() -> String { isoFmt.string(from: Date()) }

    static func sessionDir() -> URL {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("sessions")
    }

    static var sessionDirDisplayPath: String { "Files app → TabletHID → sessions" }

    // MARK: - Config builder

    private static func buildConfig(mode: DeviceMode, profileName: String, touch: TouchMouseConfig?) -> String {
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSS"
        fmt.locale = Locale(identifier: "en_US_POSIX")
        let ts = fmt.string(from: Date())

        var lines: [String] = [
            "# TabletHID Session Config",
            "# Generated: \(ts)",
            "# Mode: \(mode.title)",
            "",
            "[session]",
            "mode = \(mode.rawValue.uppercased())",
            "profile = \(profileName)",
            "generated = \(ts)",
            ""
        ]

        if let t = touch {
            lines += [
                "[touch_mouse]",
                "input_mode = \(t.mode.rawValue.uppercased())",
                "sensitivity = \(t.sensitivity)",
                ""
            ]
            lines += buttonZoneLines("left_button", b: t.leftButton)
            lines += buttonZoneLines("right_button", b: t.rightButton)
        }

        return lines.joined(separator: "\n") + "\n"
    }

    private static func buttonZoneLines(_ name: String, b: ButtonZoneConfig) -> [String] {
        [
            "[\(name)]",
            "enabled = \(b.enabled)",
            "zone_type = \(b.zoneType.rawValue)",
            "behavior = \(b.behavior.rawValue)",
            String(format: "static_left = %.3f", b.staticLeft),
            String(format: "static_top = %.3f", b.staticTop),
            String(format: "static_right = %.3f", b.staticRight),
            String(format: "static_bottom = %.3f", b.staticBottom),
            String(format: "dynamic_offset_x = %.3f", b.dynamicOffsetX),
            String(format: "dynamic_offset_y = %.3f", b.dynamicOffsetY),
            String(format: "dynamic_radius = %.3f", b.dynamicRadius),
            ""
        ]
    }
}

File: app/src/main/java/com/tablet/hid/util/SessionLogger.kt
package com.tablet.hid.util

import android.content.Context
import com.tablet.hid.model.ButtonZoneConfig
import com.tablet.hid.model.ButtonConfig
import com.tablet.hid.model.DeviceMode
import com.tablet.hid.model.GamepadConfig
import com.tablet.hid.model.JoystickConfig
import com.tablet.hid.model.TouchMouseConfig
import java.io.BufferedWriter
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.Executors

/**
 * Creates a .config + .log file pair at session start.
 * Config is a one-shot INI-style snapshot of all parameters.
 * Log receives one timestamped line per HID report via a background thread.
 */
class SessionLogger(
    context: Context,
    mode: DeviceMode,
    profileName: String,
    touchConfig: TouchMouseConfig?,
    gamepadConfig: GamepadConfig?,
) {
    private val executor = Executors.newSingleThreadExecutor { r ->
        Thread(r, "hid-logger")
    }
    private val writer: BufferedWriter
    private val isoFmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS", Locale.US)

    @Volatile private var closed = false

    init {
        val dir = LoggingStore.sessionDir(context).also { it.mkdirs() }
        val stamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())
        val stem = "${stamp}_${mode.name.lowercase()}"

        File(dir, "$stem.config").bufferedWriter().use { w ->
            w.write(buildConfig(mode, profileName, touchConfig, gamepadConfig))
        }

        writer = File(dir, "$stem.log").bufferedWriter()
        val ts = now()
        executor.execute { writer.write("$ts SESSION_START mode=${mode.name}\n") }
    }

    fun logMouse(buttons: Int, dx: Int, dy: Int, wheel: Int) {
        if (closed) return
        val ts = now()
        executor.execute {
            writer.write("$ts MOUSE buttons=$buttons dx=$dx dy=$dy wheel=$wheel\n")
        }
    }

    fun logGamepad(lx: Int, ly: Int, rx: Int, ry: Int, lt: Int, rt: Int, buttons: Int, hat: Int) {
        if (closed) return
        val ts = now()
        executor.execute {
            writer.write("$ts GAMEPAD lx=$lx ly=$ly rx=$rx ry=$ry lt=$lt rt=$rt buttons=$buttons hat=$hat\n")
        }
    }

    fun close() {
        if (closed) return
        closed = true
        val ts = now()
        executor.execute {
            try {
                writer.write("$ts SESSION_END\n")
                writer.flush()
                writer.close()
            } catch (_: Exception) {}
        }
        executor.shutdown()
    }

    private fun now(): String = isoFmt.format(Date())

    // ── Config file builder ──────────────────────────────────────────────────

    private fun buildConfig(
        mode: DeviceMode,
        profileName: String,
        touch: TouchMouseConfig?,
        gamepad: GamepadConfig?,
    ) = buildString {
        val now = now()
        appendLine("# TabletHID Session Config")
        appendLine("# Generated: $now")
        appendLine("# Mode: ${mode.displayName}")
        appendLine()
        appendLine("[session]")
        appendLine("mode = ${mode.name}")
        appendLine("profile = $profileName")
        appendLine("generated = $now")
        appendLine()

        if (touch != null) {
            appendLine("[touch_mouse]")
            appendLine("input_mode = ${touch.mode.name}")
            appendLine("sensitivity = ${touch.sensitivity}")
            appendLine()
            appendButtonZone("left_button", touch.leftButton)
            appendButtonZone("right_button", touch.rightButton)
        }

        if (gamepad != null) {
            appendButton("btn_a",      gamepad.btnA,      trigger = false)
            appendButton("btn_b",      gamepad.btnB,      trigger = false)
            appendButton("btn_x",      gamepad.btnX,      trigger = false)
            appendButton("btn_y",      gamepad.btnY,      trigger = false)
            appendButton("btn_lb",     gamepad.btnLb,     trigger = false)
            appendButton("btn_rb",     gamepad.btnRb,     trigger = false)
            appendButton("btn_lt",     gamepad.btnLt,     trigger = true)
            appendButton("btn_rt",     gamepad.btnRt,     trigger = true)
            appendButton("btn_back",   gamepad.btnBack,   trigger = false)
            appendButton("btn_start",  gamepad.btnStart,  trigger = false)
            appendButton("dpad_up",    gamepad.dpadUp,    trigger = false)
            appendButton("dpad_down",  gamepad.dpadDown,  trigger = false)
            appendButton("dpad_left",  gamepad.dpadLeft,  trigger = false)
            appendButton("dpad_right", gamepad.dpadRight, trigger = false)
            appendJoystick("left_joystick",  gamepad.leftJoystick)
            appendJoystick("right_joystick", gamepad.rightJoystick)
        }
    }

    private fun StringBuilder.appendButtonZone(name: String, b: ButtonZoneConfig) {
        appendLine("[$name]")
        appendLine("enabled = ${b.enabled}")
        appendLine("zone_type = ${b.zoneType.name}")
        appendLine("behavior = ${b.behavior.name}")
        appendLine("static_left = ${"%.3f".format(b.staticLeft)}")
        appendLine("static_top = ${"%.3f".format(b.staticTop)}")
        appendLine("static_right = ${"%.3f".format(b.staticRight)}")
        appendLine("static_bottom = ${"%.3f".format(b.staticBottom)}")
        appendLine("dynamic_offset_x = ${"%.3f".format(b.dynamicOffsetX)}")
        appendLine("dynamic_offset_y = ${"%.3f".format(b.dynamicOffsetY)}")
        appendLine("dynamic_radius = ${"%.3f".format(b.dynamicRadius)}")
        appendLine()
    }

    private fun StringBuilder.appendButton(name: String, b: ButtonConfig, trigger: Boolean) {
        appendLine("[$name]")
        appendLine("enabled = ${b.enabled}")
        appendLine("behavior = ${b.behavior.name}")
        appendLine("turbo = ${b.turbo}")
        appendLine("turbo_duration_ms = ${b.turboDurationMs}")
        appendLine("turbo_interval_ms = ${b.turboIntervalMs}")
        if (trigger) {
            appendLine("trigger_travel_dp = ${"%.1f".format(b.triggerTravelDp)}")
            appendLine("trigger_axis = ${b.triggerAxis.name}")
        }
        appendLine("offset_x = ${"%.2f".format(b.offsetX)}")
        appendLine("offset_y = ${"%.2f".format(b.offsetY)}")
        appendLine("scale_x = ${"%.3f".format(b.scaleX)}")
        appendLine("scale_y = ${"%.3f".format(b.scaleY)}")
        appendLine()
    }

    private fun StringBuilder.appendJoystick(name: String, j: JoystickConfig) {
        appendLine("[$name]")
        appendLine("enabled = ${j.enabled}")
        appendLine("deadzone = ${"%.3f".format(j.deadzone)}")
        appendLine("gain = ${"%.3f".format(j.gain)}")
        appendLine("offset_x = ${"%.2f".format(j.offsetX)}")
        appendLine("offset_y = ${"%.2f".format(j.offsetY)}")
        appendLine("scale_x = ${"%.3f".format(j.scaleX)}")
        appendLine("scale_y = ${"%.3f".format(j.scaleY)}")
        appendLine()
    }
}
